File: 0020-valid-parentheses/0020-valid-parentheses.java
class Solution {
    public boolean isValid(String s) {
            if(s.length() % 2 != 0)
                return false;
        
        Stack<Character> stack = new Stack();
        
        for(int i = 0; i< s.length();i++)
        {
            char foundChar = s.charAt(i);
            if(foundChar == '(' || foundChar == '{' || foundChar == '[')
            {
                stack.push(foundChar);
            }else{ // Closing Phase
                if(stack.isEmpty())
                {
                    return false;
                }else{
                    Character charAtPeek = stack.peek();
                    if((charAtPeek == '(' && foundChar == ')') 
                       || (charAtPeek == '{' && foundChar == '}') 
                       || (charAtPeek == '[' && foundChar == ']'))
                    {
                        stack.pop();
                    }else{
                        return false;
                    }
                }
            }
            
        }
        
        
        
        return stack.isEmpty();
        }
}

File: 0020-valid-parentheses/0020-valid-parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        if(len(s) % 2 != 0):
            return False

        stack = []

        for ch in s:
            if(ch == '(' or ch == '{' or ch == '['):
                stack.append(ch)
            else:
                if(len(stack) == 0):
                    return False
                    
                openedBracket = stack.pop()
                if((ch == ')' and openedBracket == '(') or (ch == '}' and openedBracket == '{') or (ch == ']' and openedBracket == '[')):
                    continue
                else:
                    return False
        return len(stack) == 0
            

        